lms/security.py

```diff
@@ -66,47 +67,67 @@ def get_route_policy(self, request):
         return self.fallback_policy()
 
 
-class LTISecurityPolicy:
+class _LTIUserBasedSecurityPolicy:
+    """Base class for `LTIUser` based security policies."""
+
     @classmethod
-    def authenticated_userid(cls, request):
-        if request.lti_user is None:
-            return None
+    def get_lti_user(cls, request) -> LTIUser:
+        """
+        Get and LTIUser to use as the authenticated user in this policy.
+
+        How this LTIUser is constructed is up to the subclasses.
+        """
+        raise NotImplementedError()
+
+    def permits(self, request, context, permission):
+        return _permits(self, request, context, permission)
+
+    def remember(self, request, userid, **kw):
+        pass
+
+    def forget(self, request):
+        pass
 
+    @staticmethod
+    def get_userid(lti_user):
         # urlsafe_b64encode() requires bytes, so encode the userid to bytes.
-        user_id_bytes = request.lti_user.user_id.encode()
+        user_id_bytes = lti_user.user_id.encode()
 
         safe_user_id_bytes = base64.urlsafe_b64encode(user_id_bytes)
 
         # urlsafe_b64encode() returns ASCII bytes but we need unicode, so
         # decode it.
         safe_user_id = safe_user_id_bytes.decode("ascii")
 
-        return ":".join([safe_user_id, str(request.lti_user.application_instance_id)])
+        return ":".join([safe_user_id, str(lti_user.application_instance_id)])
 
-    def identity(self, request):
-        userid = self.authenticated_userid(request)
+    def authenticated_userid(self, request):
+        identity = self.identity(request)
+        if identity is None or not identity.userid:
+            return None
 
-        if userid:
-            permissions = [Permissions.LTI_LAUNCH_ASSIGNMENT, Permissions.API]
+        return identity.userid
 
-            if any(
-                role in request.lti_user.roles.lower()
-                for role in ["administrator", "instructor", "teachingassistant"]
-            ):
-                permissions.append(Permissions.LTI_CONFIGURE_ASSIGNMENT)
+    def identity(self, request):
+        lti_user = self.get_lti_user(request)
```

Comment:
The benefits are not obvious here but now this is the only place where we need to call `get_lti_user` so doing error handling would only happen once here.